File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

struct nodeDepth {
    TreeNode *node;
    int depth;

    nodeDepth(TreeNode *n, int d) : node(n), depth(d) {}
};

class Solution {

public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        std::vector<vector<int>> result;
        if (root == nullptr) return result;
        std::queue<nodeDepth> nodeQueue;
        nodeQueue.push(nodeDepth {root, 0});

        while (!nodeQueue.empty()) {
            nodeDepth currNode = nodeQueue.front();
            nodeQueue.pop();

            if (currNode.depth >= 0 && currNode.depth < result.size()) {
                result[currNode.depth].push_back(currNode.node->val);
            }
            else {
                result.push_back({currNode.node->val});
            }
            if (currNode.node->left != nullptr) nodeQueue.push(nodeDepth {currNode.node->left, currNode.depth + 1});
            if (currNode.node->right != nullptr) nodeQueue.push(nodeDepth {currNode.node->right, currNode.depth + 1});
        } 
        return result;
    }
};

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def __init__(self):
        self.right_view = []
        self.right_depth = 0

    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        self.dfs_right(root, 1)
        return self.right_view

    def dfs_right(self, root, curr_depth):
        if root is None:
            return

        if curr_depth > self.right_depth:
            self.right_view.append(root.val)

        self.dfs_right(root.right, curr_depth + 1)
        self.dfs_right(root.left, curr_depth + 1)

        self.right_depth = max(self.right_depth, curr_depth)
        